### ISSUE TITLE
Support for zsh auto completions

### ISSUE THREAD

**Issue by KyleUltimate** [theRookieCoder/ferium#74]

**Comment by theRookieCoder** [comment]
I tried this before, and managed to get fish autocompletions working for me. The problem was that the autocompletions were given as files and I had no idea where to put the autocompletions for each shell. It was especially confusing for bash

**Comment by KyleUltimate** [comment]
for example, `paru` has zsh completions, which makes typing commands **much** easier
![image](https://user-images.githubusercontent.com/68589851/168418644-53979162-f95c-4806-8bf4-dcccc6c55fb8.png)
And for your question
https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=paru
according to `paru`(Which is an extremely popular AUR rust package) you should put them into 
![image](https://user-images.githubusercontent.com/68589851/168418712-3f17d1e9-012f-4705-8748-d6f862272f3c.png)

**Comment by theRookieCoder** [comment]
But this is only for arch linux, it's probably different for other OSs. For example `/usr/share/fish` doesn't even exist for me on macOS. And should the completions be a subcommand or should they be provided in the releases

**Comment by theRookieCoder** [comment]
Fish seems to have a [`$fish_complete_path`](https://fishshell.com/docs/current/completions.html#where-to-put-completions) and for zsh I found [this](https://unix.stackexchange.com/questions/607809/where-should-i-install-the-zsh-completion-file-as-a-package-maintainer)

**Comment by KyleUltimate** [comment]
Yeah it seems like you can put it in `/usr/local/share/zsh/site-functions`
And for my system, indeed a lot of program put their zsh's in there

**Comment by theRookieCoder** [comment]
And I think bash doesn't even support completions, it needs some sort of plugin if I remember correctly

**Comment by theRookieCoder** [comment]
Oh and FYI the completions will be generated by [clap complete](https://crates.io/crates/clap_complete)

**Comment by KyleUltimate** [comment]
> And I think bash doesn't even support completions, it needs some sort of plugin if I remember correctly

iirc, it's bash-completions? long time since I've used bash tho

**Comment by TheHackerCoding** [comment]
> But this is only for arch linux, it's probably different for other OSs. For example `/usr/share/fish` doesn't even exist for me on macOS. And should the completions be a subcommand or should they be provided in the releases?

For completions in fish for any *nix you can either put them in `.config/fish/completions` (user specific) or `/etc/fish/completions` (global)

**Comment by linuskr** [comment]
You could also let the user sort out where the files should go, and just provide a command that prints the autocompletion script for the specified shell to stdout, like `clap_complete::generator::generate`. 
This seems to be the only useable method for `cargo install` anyway, since it can't place files in arbitrary locations.

**Comment by theRookieCoder** [comment]
Yup I believe that's what packwiz does. And also the plan was to write to the proper file automatically when the subcommand is run

**Comment by imsofi** [comment]
> Yup I believe that's what packwiz does. And also the plan was to write to the proper file automatically when the subcommand is run

Id recommend to let each of the packaging systems to handle the install process of this to the correct location. As each distro does things slightly differently. (A good example of one that sticks out from the norm would be nix, which installs it to `/nix/store/...`)

For cargo based installs, using a command similar to `ferium hook bash`/`ferium hook zsh` to print the shell script to stdout would be the way to go. As then end users would only need to add `eval "$(ferium hook bash)"` to their `.bashrc` file for example. Which should be a good compromise for the standalone cargo install.

**Comment by theRookieCoder** [comment]
I've used the print to stdout approach. I believe package managers can easily write to the appropriate files this way, without the need to maintain completion files

**Comment by KyleUltimate** [comment]
All AUR releases has been updated to use it!
It *should* work

**Comment by SettingDust** [comment]
Why `ferium complete PowerShell` shows `Invalid value "PowerShell" for '<SHELL>': Invalid variant: PowerShell`

There is in clap. https://docs.rs/clap_complete/latest/clap_complete/shells/enum.Shell.html

**Comment by theRookieCoder** [comment]
You have to provide the argument in lower case, so `ferium complete powershell`. Just a quirk in the way clap arg enums work I guess

**Comment by SettingDust** [comment]
> ferium complete powershell

okay
